framework/src/base/MooseApp.C

```diff
@@ -358,10 +364,22 @@ MooseApp::setInputFileName(std::string input_filename)
   _input_filename = input_filename;
 }
 
+const MooseApp &
+MooseApp::root() const
+{
+  if (!parent())
+    return *this;
+  return parent()->root();
+}
+
 std::string
-MooseApp::getOutputFileBase()
+MooseApp::levelString()
 {
-  return _output_file_base;
+  if (isUltimateMaster())
+    return "";
+  std::string base = root().getFileName();
```

Comment:
Dont "pull" like this.  _Push_ information down the hierarchy.  It's like this on purpose.